### PR TITLE
Introduce Server KeyManager Facade Interface

### PR DESCRIPTION
**Pull request by azdagron**
This is the next facade introduced as part of #2153.

It is a little meatier than other PRs, but took the opportunity to clean up the keymanager base tests and implement them in terms of the facade interface. Most of the v0 shim implementation was a slight rework of the helper functions pulled from the cryptoutil and x509util packages. The conformance of the v0 shim is tested exhaustively as part of the keymanager base tests. I've punted on more exhaustive tests on the shim until I introduce the v1 shim. I think there is going to be opportunity to do something fancy here and I don't want to have to undo things to fit it in.

### CODE REVIEW

File: pkg/common/telemetry/server/keymanager/wrapper.go [hunk 1]
```diff
@@ -4,44 +4,41 @@ import (
...
 		k: km,
 	}
 }
 
-func (w serverKeyManagerWrapper) GenerateKey(ctx context.Context, req *keymanager.GenerateKeyRequest) (_ *keymanager.GenerateKeyResponse, err error) {
+func (w serverKeyManagerWrapper) GenerateKey(ctx context.Context, req *keymanagerv0.GenerateKeyRequest) (_ *keymanagerv0.GenerateKeyResponse, err error) {
 	callCounter := StartGenerateKeyCall(w.m)
 	defer callCounter.Done(&err)
 
-	ctx, cancel := context.WithTimeout(ctx, keymanager.RPCTimeout)
```
Comment (azdagron) [inline pkg/common/telemetry/server/keymanager/wrapper.go:26]: This didn't belong here.

File: pkg/common/x509util/cert.go [hunk 2]
```diff
@@ -18,31 +15,6 @@ func CertificateMatchesPrivateKey(certificate *x509.Certificate, privateKey cryp
 	return cryptoutil.KeyMatches(privateKey, certificate.PublicKey)
 }
 
-func CreateCertificate(ctx context.Context, km keymanager.KeyManager, template, parent *x509.Certificate, parentKeyID string, publicKey crypto.PublicKey) (*x509.Certificate, error) {
```
Comment (azdagron) [inline pkg/common/x509util/cert.go:21]: This functionality is now implemented by the shim.

File: pkg/server/catalog/catalog.go [hunk 3]
```diff
@@ -108,11 +109,6 @@ func BuiltIns() []catalog.Plugin {
 	return append([]catalog.Plugin(nil), builtIns...)
 }
 
-type DataStore struct {
```
Comment (azdagron) [inline pkg/server/catalog/catalog.go:111]: Opportunistic cleanup. This isn't needed.

File: pkg/server/endpoints/bundle/acme_auth.go [hunk 4]
```diff
@@ -123,32 +118,17 @@ func (ks *acmeKeyStore) NewPrivateKey(ctx context.Context, id string, keyType au
...
-		kmKeyType = keymanager.KeyType_RSA_2048
+		kmKeyType = keymanager.RSA2048
 	case autocert.EC256:
-		kmKeyType = keymanager.KeyType_EC_P256
+		kmKeyType = keymanager.ECP256
 	default:
 		return nil, errs.New("unsupported key type: %d", keyType)
 	}
 
-	ctx, cancel := context.WithTimeout(ctx, keymanager.RPCTimeout)
```
Comment (azdagron) [inline pkg/server/endpoints/bundle/acme_auth.go:133]: Shim applies this to all RPCs.

File: pkg/server/plugin/keymanager/base/keymanagerbase.go [hunk 5]
```diff
@@ -12,51 +12,78 @@ import (
...
+	*keymanagerv0.PublicKey
 }
 
-type ErrorFn func(format string, args ...interface{}) error
-type WriteFn func(ctx context.Context, entries []*KeyEntry) error
-
-type Impl struct {
-	ErrorFn ErrorFn
-	WriteFn WriteFn
+// Funcs is a collection optional callbacks. Default implementations will be
```
Comment (amartinezfayo) [inline pkg/server/plugin/keymanager/base/keymanagerbase.go:27]:
```suggestion
// Funcs is a collection of optional callbacks. Default implementations will be
```

File: pkg/server/plugin/keymanager/v0.go [hunk 6]
```diff
@@ -0,0 +1,156 @@
...
+		return nil, v0.Errorf(codes.Internal, "plugin response missing public key for public key %q", id)
+	case pb.Id != id:
+		return nil, v0.Errorf(codes.Internal, "plugin response has unexpected key id %q for public key %q", pb.Id, id)
+	case len(pb.PkixData) == 0:
+		return nil, v0.Errorf(codes.Internal, "plugin response missing public key PKIX data for public key %q", id)
+	}
+
+	publicKey, err := x509.ParsePKIXPublicKey(pb.PkixData)
+	if err != nil {
+		return nil, v0.Errorf(codes.Internal, "unable to parse public key pkix data for public key %q: %v", id, err)
```
Comment (amartinezfayo) [inline pkg/server/plugin/keymanager/v0.go:85]: nit: consistent casing
```suggestion
		return nil, v0.Errorf(codes.Internal, "unable to parse public key PKIX data for public key %q: %v", id, err)
```

**Review by amartinezfayo — COMMENTED** [review]
Looks great @azdagron!
I only have a couple of very small nits.

**Review by amartinezfayo — APPROVED** [review]

**Pull request merged**